workers/price-reporter/src/manager/external_executor.rs

```diff
@@ -326,34 +331,39 @@ async fn ws_handler_loop(
                 // Forward incoming messages from the external price reporter
                 // to the executor
                 Some(res) = ws_read.next() => {
-                    if let Err(e) = handle_incoming_ws_message(res, &subscription_states, &msg_in_tx).await {
-                        error!("Error handling incoming message from external price reporter: {e}");
+                    if let Err(e) = handle_incoming_ws_message(res, &msg_in_tx) {
+                        error!("Error handling incoming message from external price reporter: {e}, retrying...");
                         break;
                     }
                 }
 
                 // Forward outgoing messages from the executor to the external price reporter
                 Some(message) = msg_out_rx.recv() => {
                     if let Err(e) = ws_write.send(Message::Text(serde_json::to_string(&message).unwrap())).await {
-                        error!("Error sending message to external price reporter: {e}");
+                        error!("Error sending message to external price reporter: {e}, retrying...");
                         break;
                     }
                 },
             }
         }
+
+        // Breaking from the loop above indicates a failure in the websocket connection.
+        // As such, we will have to re-subscribe to all the price streams that
+        // were previously subscribed to on the re-established connection, so we
+        // enqueue the re-subscription jobs here
+        info!("Enqueueing re-subscriptions to prior price streams...");
+        resubscribe_to_prior_streams(&price_reporter_job_queue, &subscription_states).await?;
```

Comment:
Why not just have the already existing retry logic re-subscribe? Feels clunky to go through the job queue when we already have retry logic executed immediate after this (i.e. at the top of the next loop).

Feel like we can just pass `subscription_states` to `connect_with_retries` (maybe rename this to `connect_and_resubscribe`?

Reply:
Oh, haha - I did this because I thought it was cleaner than the alternative 😅 guess we shot past each other on this one.

I err against doing resubscriptions in the `connect_with_retries` method because on the first pass of the loop, we would double-subscribe to all price streams. this is because the `StreamPrice` jobs from the handshake manager are typically received before the `ws_handler_loop` is spawned (I have observed this), the outgoing subscription messages on the websocket are already enqueued & the subscription states map written to by the time the websocket is established. currently, the external price reporter is configured to handle double-subscriptions idempotently, but this doesn't feel like an assumption that should be made here.

As far as using the job queue vs the outbound websocket message channel directly, we already have the logic for subscribing to a new price stream well-defined in the handler for the price stream subscription job, and that generally is the interface for new subscriptions, so why not use it?

But I digress, I'm not deeply opinionated, and if you feel this violates patterns that have already been established in the codebase, I'll defer to consistency

Reply:
Something else I considered when deciding on this approach: when implementing stablecoin-quote price conversion, I'll want the following logic:
If we get a `StreamPrice` job for e.g. UNI-USDC, we'll want to ensure that we are also subscribed to both UNI-USDT and USDC-USDT for the conversion.
I was planning on just sending `StreamPrice` jobs for these additional pairs to the job queue, similarly to how we do here. I feel this centralizes & simplifies the means by which we subscribe to price streams instead of having yet another code path responsible for this

Reply:
The fact that we're designing this with the implementation of the API in mind means the API interface is bad. And playing games like "which message arrives first" is not a principled approach to design.

Way more readable to run a method than go through the queue

Reply:
yeah, fair points. I'll go ahead and restructure w/o the queue